torch/cuda/__init__.py

```diff
@@ -299,9 +299,93 @@ def get_device_properties(device: _device_t) -> _CudaDeviceProperties:
         raise AssertionError("Invalid device id")
     return _get_device_properties(device)  # type: ignore[name-defined]
 
+def get_current_device_index():
+    if torch.cuda._cuda_getDeviceCount() > 0:
+        return torch.cuda._cuda_getDevice()
+    return -1
+
+def get_device_index(device: Optional[_device] = None, optional: bool = False, allow_cpu: bool = False) -> int:
+    if device is None:
+        if optional:
+            return get_current_device_index()
+        else:
+            raise ValueError('Expected a torch.device with a specified index '
+                             'or an integer, but got:{}'.format(device))
+    device_index = -1
+    if isinstance(device, str):
+        device = torch.device(device)
+
+    if isinstance(device, torch.device):
+        if not allow_cpu and device.type == 'cpu':
+            raise ValueError('Expected a non cpu device, but got: {}'.format(device))
+        device_index = -1 if device.type == 'cpu' else torch.cuda._cuda_getDeviceIndex(device)
+
+    if isinstance(device, int):
+        device_index = device
+
+    return device_index
+
+class device_jit(object):
+    r"""Context-manager that changes the selected device.
+    This is similar to device (torch.device or int), but has been
+    introduced for jit compatibility
+
+    Arguments:
+        device_jit (torch.device or int): device index to select. It's a no-op if
+            this argument is a negative integer or ``None``.
+    """
 
-@contextlib.contextmanager
-def stream(stream):
+    def __init__(self, device: _device):
+        self.idx = get_device_index(device)
+        self.prev_idx = -1
+
+    def __enter__(self):
+        if self.idx == -1:
+            return
+        self.prev_idx = torch.cuda._cuda_getDevice()
+        if self.prev_idx != self.idx:
+            torch.cuda._cuda_setDevice(self.idx)
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        if self.prev_idx != self.idx:
+            torch.cuda._cuda_setDevice(self.prev_idx)
+
+class get_current_stream(object):
```

Comment:
```suggestion
class StreamContext(object):
```
Let's try to come up with a more accurate name for this. It is doing more than getting the current stream.